tensorboard/webapp/runs/store/runs_selectors.ts

```diff
@@ -85,16 +85,65 @@ export const getRuns = createSelector(
   }
 );
 
+/**
+ * Determines hparam data for each run in the active route.
+ *
+ * Attempts to match each run with a session and, if found, copies the hparam
+ * values from the corresponding session group to the run.
+ *
+ * Note, it returns an RunToHparamsAndMetrics but leaves the `metrics` field
+ * blank since the Hparams data sources does not actually retrieve metrics
+ * data.
+ *
+ * Meant for usage in the Dashboard views.
+ */
+export const getDashboardRunsToHparams = createSelector(
+  getDashboardSessionGroups,
+  getExperimentIdsFromRoute,
+  getDataState,
+  (dashboardSessionGroups, experimentIds, state): RunToHparamsAndMetrics => {
+    if (!experimentIds) {
+      return {};
+    }
+
+    const runIds: string[] = [];
+    for (const experimentId of experimentIds) {
+      runIds.push(...(state.runIds[experimentId] || []));
+    }
+
+    const runToHparamsAndMetrics: RunToHparamsAndMetrics = {};
+    for (const sessionGroup of dashboardSessionGroups) {
+      const hparams: HparamValue[] = Object.entries(sessionGroup.hparams).map(
+        (keyValue) => {
+          const [hparam, value] = keyValue;
+          return {name: hparam, value};
+        }
+      );
+
+      for (const session of sessionGroup.sessions) {
+        for (const runId of runIds) {
+          if (runId.startsWith(session.name)) {
+            runToHparamsAndMetrics[runId] = {
+              hparams,
+              // The underlying data source that fetches the session groups data
+              // does not retrieve metrics.
+              metrics: [],
+            };
+          }
+        }
+      }
```

Comment:
nit:
I know performance on the client isn't always the biggest deal, so you can probably ignore this.

There are experiments with 1000s of runs and potentially a large number of sessions this could get kinda crazy.

If you flattened out the sessions to a `Record<session.name, HparamValue[]>` you could then iterate over the `runIds` a single time.

Reply:
More or less, here is approximately what I was thinking written out in TS (I didn't run it but I THINK this works).
```
const hparamsBySessionName = Object.fromEntries(dashboardSessionGroups
    .map((sessionGroup) => {
        const hparams: HparamValue[] = Object.entries(sessionGroup.hparams).map(
            (keyValue) => {
              const [hparam, value] = keyValue;
              return {name: hparam, value};
            });
        return sessionGroup.sessions.map((session) => [session.name, hparams])
    }).flat());

const sessionNames = Object.keys(hparamsBySessionName).sort((a, b) => b.length - a.length);

for (const runId of runIds) {
    const sessionName = sessionNames
        .find((sessionName) => sessionName.startsWith(runId));
    runToHparamsAndMetrics[runId] = {
        hparams: hparamsBySessionName[sessionName].hparams,
        metrics: [],
    };
}
```

While writing this I did wonder about an edgecase though.

Imagine I have two different runs say 'run1' and 'run10'. Your initial solution will set the hparams of 'run10' to whichever corresponding session group comes LATER. Your re-implementations will set it to whichever comes FIRST.

By sorting the `sessionNames` by length descending I BELIEVE this can always be avoided.

Reply:
I implemented the change although I hesitated. Some considerations:

Against:
* your suggestion still has the same runtime complexity O(r*s) (where r is number of runs and s is number of sessions).
* it has ~ the same worst-case runtime, and it's not hard to imagine real-world scenarios where it will hit close to worst case.  

For:
* the changes are nonetheless more efficient in practice because of the early return with "break;" (or, in your code sample, with "find()"). On average, it should reduce the number of comparisons by half.
* the changes aren't overly complex and I don't think pass the threshold into "premature optimization" territory.

On the other hand, if we really cared about making this performant for 1000s or 10000s of runs then we'd probably want to consider an entirely different method. Maybe a trie (prefix tree) for storing the run names.

Reply:
I hesitated to even make this suggestion because I was aware that they didn't change the overall time complexity. I actually did consider suggesting a prefix tree but thought the complexity outweighed the realized performance.

Reply:
At first I ignored the comments about session name order but offline you convinced me that it was a serious problem. I added a test to illustrate and made the fix.

Reply:
Thanks for addressing the issue.